Merge upstream 1.21.0 into our fork

tilo opened the pull request:
merging upstream/master @1.20.0 into our fork, so we can get rid of our private fork of this gem.

this also includes the changes which were pushed upstream to use HashWithIndifferentAccess
https://github.com/JsonApiClient/json_api_client/pull/386
and
https://github.com/JsonApiClient/json_api_client/pull/389

coreyward committed:
Exclude path params from attributes hash

coreyward committed:
Test path params are excluded from serialized attributes

mtsmfm committed:
Consider `Model.page(nil)`

It is very happy if we can write `Model.page(params[:page])` like Kaminari.

mordaroso committed:
Fix custom type comment to use TypeFactory

senid231 committed:
#278 updating faraday to fix issue when param value is an empty array

senid231 committed:
allow to send fields and/or includes on create/update resource

senid231 committed:
Merge pull request #286 from senid231/coreyward-f/serialized-attribs

Exclude path params from attributes hash

senid231 committed:
Merge pull request #287 from senid231/github-mtsmfm-page-nil

allow nil for page number parameter

senid231 committed:
Merge pull request #282 from mordaroso/patch-1

Fix custom type comment to use TypeFactory

senid231 committed:
Merge pull request #284 from senid231/278-update-faraday-to-fix-empty-array-param-issue

#278 updating faraday to fix issue when param value is an empty array

senid231 committed:
Merge pull request #285 from senid231/query-params-on-create-and-update-resource

allow to send fields and/or includes on create/update resource

gaorlov committed:
adding test that shows the error: chaining scopes modifies the parent scope

gaorlov committed:
fixing readme to update test and coverage icon targets

gaorlov committed:
fixing sopes to be immutable

gaorlov committed:
fixing paginator to accept strings or symbols as paer_page key

gaorlov committed:
adding test that demonstrates Query::Builder#find with args doesn't work

gaorlov committed:
fixing Query::Builder#find with parameters

gaorlov committed:
Merge pull request #293 from gaorlov/idempotent_scoping

#292 #254 making scopes unmodifiable

gaorlov committed:
version bump v1.6.0

edymerchk committed:
Use mocha/minitest instead of mocha/mini_test on test_helper

fusion2004 committed:
update arguments for custom connections run method

senid231 committed:
Merge pull request #297 from edymerchk/fix-deprecation-warning

Use mocha/minitest instead of mocha/mini_test on test_helper

senid231 committed:
Merge pull request #298 from fusion2004/update-readme-for-new-custom-connection-arguments

Update arguments for custom connections run method in README

jsmestad committed:
Merge pull request #295 from gaorlov/coverage_readme_fix

fixing readme to update test and coverage icon targets

senid231 committed:
#303 optional add default to changes

senid231 committed:
cache values for defined properties and relationships [performance optimization]

senid231 committed:
#299 optional get relationship from dataset

senid231 committed:
#291 import README - pagination override examples

senid231 committed:
#290 symbolize params keys on model initialize

it will fix behavior when relationships and links are passed as symbol keys

senid231 committed:
Merge pull request #306 from senid231/291-improve-readme-pagination-override-example

#291 import README - pagination override examples

senid231 committed:
Merge pull request #307 from senid231/290-new-model-allow-symbolized-relationship-key

#290 symbolize params keys on model initialize

senid231 committed:
Merge pull request #304 from senid231/303-optional-add-default-to-changes

#303 optional add default to changes

senid231 committed:
Merge pull request #300 from senid231/smart-performance-optimization

define methods for properties and associations getter/setter

gaorlov committed:
version bump 1.6.1

gaorlov committed:
version bump 1.6.1

senid231 committed:
Merge pull request #310 from senid231/bump_161

version bump 1.6.1

jamesstonehill committed:
Raise client error for unhandled 4xx responses

jamesstonehill committed:
Add PR #311 to CHANGELOG

gaorlov committed:
Merge pull request #311 from jamesstonehill/raise-on-misc-client-error

Raise client error for unhandled 4xx responses

gaorlov committed:
resolving diff with upstream

gaorlov committed:
version bump v1.6.2

gaorlov committed:
updating chantgelog

jamesstonehill committed:
Do not raise errors on 422 responses

JSON APIs often return validation errors with 422 so the client would be
notified that the response was not successful by the existence of the
resource errors.

senid231 committed:
Merge pull request #312 from jamesstonehill/422-error-handling

Do not raise errors on 422 responses

gaorlov committed:
verion bump v1.6.3

gaorlov committed:
fixing typo in changelog

hashrocketeer committed:
Mimics ActiveRecord behavior when destroying resource

* Add `destroyed?` method
* Return false on `persisted?`
* Do not clear resource attributes

Co-authored-by: Gabriel Reis <gabriel@hashrocket.com>

hashrocketeer committed:
Do not clear relationships on destroy

Co-authored-by: Mary Lee <mary.lee@hashrocket.com>
Co-authored-by: Gabriel Reis <gabriel@hashrocket.com>

senid231 committed:
add shallow path for belongs_to

decouple attributes and path params

fix #180

gaorlov committed:
Merge pull request #314 from hashrocketeer/mimic_destroy

Mimics ActiveRecord behavior when destroying resource

gaorlov committed:
readme update; changelog update; version bump v1.6.4

fzf committed:
Allow custom error messages

* Adds ability to override key in resource to surface different error messages

senid231 committed:
Merge pull request #315 from senid231/180-add-shallow-path-for-belongs-to

add shallow path for belongs_to

senid231 committed:
add tests for update resource for correct relationships dirty behavior

ensures fetched relationships will not be added to payload

senid231 committed:
Merge pull request #318 from senid231/add-test-for-update-resource

add tests for update resource for correct relationships dirty behavior

senid231 committed:
290 fix passing relationships on create

also fix false positive tests

senid231 committed:
Merge pull request #320 from senid231/290-fix-passing-relationship-on-create

#290 fix passing relationships on create

senid231 committed:
add missing changelog

add #315 and #320 to unreleased section of CHANGELOG.md

gaorlov committed:
Merge pull request #321 from senid231/fix-changelogs

add missing changelog

gaorlov committed:
version bump: 1.7.0

Commit:
Retain context from Faraday errors

senid231 committed:
Merge pull request #322 from COzero/retain-error-context

Retain context from Faraday errors

senid231 committed:
add status_handlers to JsonApiClient::Resource.connection_options

it allows to override status handling

senid231 committed:
fix travis for ruby 2.2

latest bundler drop support of ruby 2.2
we will install bundler < 2.0 for ruby < 2.3

senid231 committed:
Merge pull request #325 from senid231/fix-travis-for-ruby-2.2

fix travis for ruby 2.2

senid231 committed:
correct changes after initialize resource

remove type from changes on initialize
ensure that query builder doesn't propagate query values to resource attributes via #build method

senid231 committed:
Merge pull request #316 from 1debit/customize_errors

Allow custom error messages

senid231 committed:
add #316 to changelog

senid231 committed:
Merge pull request #327 from senid231/update-changelog

add #316 to changelog

senid231 committed:
allow custom type for models

fix #288

senid231 committed:
Merge pull request #305 from senid231/299-get-included-from-dataset

#299 optional get relationship from dataset

gaorlov committed:
version bump: 1.8.0

gaorlov committed:
adding test for destruction with custom primary key

gaorlov committed:
allowing resources to be deleted with a custom primary key

senid231 committed:
Merge pull request #328 from senid231/allow-custom-type

allow custom type for models

senid231 committed:
Merge pull request #326 from senid231/ensure-correct-changes-on-resource-init

correct changes after initialize resource

senid231 committed:
changelog fix after 1.8.0 release

senid231 committed:
Merge pull request #331 from senid231/changelog-fix-after-1.8.0

changelog fix after 1.8.0 release

senid231 committed:
Merge pull request #324 from senid231/custom-status-handling

add possibility to override status handling

senid231 committed:
Merge pull request #330 from gaorlov/primary_key_fix

Deletion for primary keys

senid231 committed:
upgrade changelog (add #330)

senid231 committed:
Merge pull request #332 from senid231/changelog-upgrade

upgrade changelog

gaorlov committed:
version bump: 1.9.0

senid231 committed:
add failing test that reproduces issue #334

senid231 committed:
retrieve assigned relationship record via association getter method

with caching

senid231 committed:
fix byebug version fo travis ci

senid231 committed:
Merge pull request #335 from senid231/334-access-to-assigned-relationship

access to assigned relationship

ChrisBr committed:
Implement eql and hash for Builder class

gaorlov committed:
version bump: 1.10.0

senid231 committed:
update changelog for version 1.10.0

stokarenko committed:
Add `raise_on_blank_find_param` resource setting

gaorlov committed:
Merge pull request #344 from stokarenko/add-raise_on_blank_find_param-resource-setting

Add `raise_on_blank_find_param` resource setting

gaorlov committed:
version bump: v1.11.0

stokarenko committed:
Track the real HTTP reason of ApiErrors

lcpriest committed:
Add an immutable option to JSON API resources.

Discussed here:
https://github.com/JsonApiClient/json_api_client/issues/343

Commit:
chore: update .gitignore to ignore IntelliJ and asdf files

gaorlov committed:
Merge pull request #345 from stokarenko/better-api-errors

Track the real HTTP reason of ApiErrors

gaorlov committed:
version bump: v 1.12.0

Commit:
feat: add NestedParamPaginator to make pagination params consistent

fixes #347

stokarenko committed:
Fix resource including for STI objects

gaorlov committed:
Merge pull request #349 from stokarenko/fix-included-data-for-sti

Fix resource including for STI objects

gaorlov committed:
version bump: v1.12.1

stokarenko committed:
Fix including with blank relationships

gaorlov committed:
Merge pull request #350 from stokarenko/fix-including-with-blank-relationships

Fix including with blank relationships

gaorlov committed:
version bump: v1.12.2

Commit:
chore: add NestedParamPaginator to README and CHANGELOG

gaorlov committed:
Merge branch 'master' into feature/fix_pagination_issues

stokarenko committed:
Fix resource serialization involving last_result_set

gaorlov committed:
Merge pull request #348 from dugsmith/feature/fix_pagination_issues

fix: make page[] param wrap consistent

gaorlov committed:
version bump: v1.13.0; readme update

gaorlov committed:
Merge branch 'master' into fix-resource-serialization-involving-last-result-set

gaorlov committed:
Merge branch 'master' into update-changelog-after-1-10

gaorlov committed:
Merge pull request #341 from senid231/update-changelog-after-1-10

update changelog for version 1.10.0

senid231 committed:
Merge pull request #351 from stokarenko/fix-resource-serialization-involving-last-result-set

Fix resource serialization involving last_result_set

senid231 committed:
Merge branch 'master' into feature/implement_hash

gaorlov committed:
Merge pull request #338 from ChrisBr/feature/implement_hash

Implement eql and hash for Builder class

gaorlov committed:
version bump: v1.14.0

Commit:
test: show relationships w/o included error (issue #352)

Commit:
fix: resolve issue #352

gaorlov committed:
Merge pull request #353 from dugsmith/fix_included_data

fix issue #352: relationships w/o included data

gaorlov committed:
version bump: 1.14.1

gaorlov committed:
adding gemfile for rails 5.2.3; adding ruby 2.5.1, 2.6.0; trying to dd ruby-head

gaorlov committed:
removing incompatible ruby-rails versions from travis

gaorlov committed:
removing incompatible ruby-rails versions from travis

gaorlov committed:
Merge pull request #355 from gaorlov/travis-updates

Adding test support for rubies 2.5.1, 2.6.0, ruby-head; rails 5.2.3

yahonda committed:
Add gemfile for Rails 6.0

* The latest version of Rails 6.0 is Rails 6.0.0.rc2
* Rails 6.0 requires Ruby 2.5.0 then excluded CI agains Ruby 2.4 or
lower

Refer Rails 6.0 required_ruby_version
https://github.com/rails/rails/blob/16f2cd88e3ee92de341ba30b102e9206be3c68a7/rails.gemspec#L12

yahonda committed:
Support Rails 6.0.0

senid231 committed:
fix missing constant formatter

https://github.com/JsonApiClient/json_api_client/issues/333

ericotto committed:
Fix indent in README

gaorlov committed:
Merge pull request #358 from ericotto/master

Fix indent in README

gaorlov committed:
Merge pull request #356 from yahonda/ci_against_rails60

Add gemfile for Rails 6.0

gaorlov committed:
Merge pull request #357 from senid231/fix-missing-constant-formatter

fix missing constant formatter

gaorlov committed:
Merge pull request #346 from lcpriest/feature/readonly-resource

Add an immutable option to JSON API resources.

gaorlov committed:
updating changelog; removing ruby-head from travis matrix

gaorlov committed:
version bump: v1.15.0

gaorlov committed:
Merge branch 'master' of github.com:JsonApiClient/json_api_client

stokarenko committed:
Support gzip content encoding

gaorlov committed:
Merge pull request #359 from stokarenko/gzip-support

Support gzip content encoding

gaorlov committed:
version bump: v1.16.0

Commit:
call from method so that it will execute parent classes' implementation

Commit:
update CHANGELOG

gaorlov committed:
Merge pull request #361 from mqchau/callSuperFromInheritedMethod

call super from inherited method so that it will execute parent classes' implementations

gaorlov committed:
version bump: v1.16.1

sikachu committed:
Relax faraday and faraday_middleware version

gaorlov committed:
Merge pull request #364 from sikachu/relax-faraday-versions

Relax faraday and faraday_middleware version

gaorlov committed:
version bump: v1.17.0

dpep committed:
bigdecimal 2

BigDecimal 2.0 no longer supports `.new`.  looks like this gem uses both `BigDecimal(x)` and `BigDecimal.new(x)` - mind if we upgrade these last few calls?

dpep committed:
updating travis matrix

dpep committed:
fix warnings

Ruby 2.7 deprecates automatic conversion from a hash to keyword arguments
https://blog.saeloun.com/2019/10/07/ruby-2-7-keyword-arguments-redesign.html

fixing those up, plus one more I noticed in the logs from https://github.com/JsonApiClient/json_api_client/pull/370#issuecomment-683349721

https://travis-ci.org/github/JsonApiClient/json_api_client/jobs/722402404

thanks @sharshenov

gaorlov committed:
Merge pull request #371 from dpep/fix-warnings

fix warnings

gaorlov committed:
Merge pull request #370 from dpep/bigdecimal-2

bigdecimal 2 support

inderps committed:
Fix: Cannot access associations of 'dashed' type for a newly initialised resource.

inderps committed:
#372 - Fix handling of dashed-types associations correctly

gaorlov committed:
Merge pull request #372 from inderps/fix/initializing-resources-with-associations-of-dashed-types

Fix: Cannot access associations of 'dashed' type for a newly initialised resource.

gaorlov committed:
version bump: v1.18.0

mtoygar committed:
fix broken links on CHANGELOG.md

Commit:
merging upstream 1.18.0 into master

corsonknowles commented:
I untagged myself as a reviewer. Perhaps someone from @1debit/disputes  or @1debit/financial-health can pitch in!

tilo commented on lib/json_api_client/resource.rb:
```diff
@@ -310,21 +354,23 @@ def _build_connection(rebuild = false)
     # @param params [Hash] Attributes, links, and relationships
     def initialize(params = {})
-      params = params.with_indifferent_access
+      params = params.with_indifferent_access # upstream has: params = params.symbolize_keys
```
- do we really need access via strings?

- the new upstream versions use `symbolize_keys`

ldrbrandon replied:
Oof, I think this will have some pretty serious implications if we switch to using keys. Likely we'll have to update in several places; which is fine if we're committing to that, just want to point it out

tilo replied:
@ldrbrandon yes, that's why I left our code in place for now :(

tilo replied:
https://github.com/rails/rails/blob/cb0c746b7d757d8491f2db1539cf9377b3eb833e/activesupport/lib/active_support/core_ext/hash/indifferent_access.rb#L9

ldrbrandon commented:
Without digging into this too much it seems like a good idea but IMO, if we're going to do this, we should QA this branch on any application that would use it and ensure that it's not going to break things.

Commit:
update

Commit:
update

openbl committed:
Display CI Status of "master" Branch Only

Travis CI, without the "branch" query param, will show the latest build regardless of the branch, which is a misleading indication of the health of the repo. At the time of this commit, "master" is 🟢 while [the latest build](https://travis-ci.org/github/JsonApiClient/json_api_client/builds/768058352) is 🔴 .

emancu committed:
Add extra error classes to handle server errors.

In order to provide more granularity on the exceptions, we are adding some additional
error classes for the common Server errors.

emancu committed:
Add 408-RequestTimeout class

openbl commented on README.md:
```diff
@@ -1,16 +1,16 @@
+<<<<<<< HEAD
```
This appears to have been new [from what I can tell
](https://github.com/JsonApiClient/json_api_client/compare/1debit:merge-upstream-1.18.0#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5)

```suggestion
```

tilo replied:
oops

openbl reviewed (approved):
You only added a few doc changes so seem great. Just caught something from a merge conflict, and extraneous "<<<<<<<".

openbl commented:
Also, I said this [in chat](https://chime.slack.com/archives/C016Q5ZENHG/p1620657998329300?thread_ts=1620408473.321800&cid=C016Q5ZENHG) but will also mention here
> i wonder if we could accomplish our diff by inheriting our own resource class to do those things instead of needing to hack their library

openbl reviewed (dismissed):
Talked about this w/ @tilo in person and we began to feel less comfortable w/ merging these changes without better inter-service/integration testing since it's probably the area least covered by testing outside of perhaps what QE is doing.

Commit:
merge master into branch

Commit:
use HashWithIndifferentAccess, instead of symbolize_keys

gaorlov committed:
Merge pull request #379 from openbl/patch-1

Display CI Status of "master" Branch Only

gaorlov committed:
Merge pull request #374 from mtoygar/master

fix broken links on CHANGELOG.md

emancu committed:
Update lib/json_api_client/errors.rb

Co-authored-by: Tilo <tilo@users.noreply.github.com>

emancu committed:
Update CHANGELOG

Commit:
update CHANGELOG

Commit:
update

Commit:
merge Chime master into branch

donovan-lampa reviewed (approved):
Let's do it. 

I see up above the concerns about testing and rollout, do we have a plan for how we want to go about rolling this out or testing it? One app at a time in QA first? Is there meaningful traffic we can push to apps in QA to get good enough coverage to feel confident enough? Is our integration testing story better now (would RING cover this?)

Commit:
removing pry

gaorlov committed:
Merge pull request #386 from tilo/HashWithIndifferentAccess

use HashWithIndifferentAccess, instead of symbolize_keys

gaorlov committed:
Merge pull request #382 from carwow/add_extra_error_classes

Add extra error classes to handle server errors.

gaorlov committed:
version bump: v1.19.0

tilo commented:
@donovan-lampa I'll update it once more with the recent updates to the upstream - so the HashWithIndifferentAccess is also included.

Commit:
merging 1.19.0 upstream into branch

tilo commented:
@corsonknowles @ldrbrandon @dpep @donovan-lampa @openbl @dalyons 

Please note that I created a new branch "upstream" in our fork which forks off of 3f36f74, which contans all the upstream changes since we created our fork with our own modifications.

This makes it easier to see our actual diff:

https://github.com/1debit/json_api_client/compare/upstream...merge-upstream-1.18.0

tilo commented:
I'll create more PRs for the upstream to bring our changes in, e.g. the bang methods, to make our diff even smaller

https://github.com/JsonApiClient/json_api_client/pull/389

Commit:
adding bang-methods for create and update

Commit:
adding tests

Commit:
update CHANGELOG.md

gaorlov committed:
Merge pull request #389 from tilo/create-bang-methods

adding create!, update_attributes!, update! methods

gaorlov committed:
version bump: 1.20.0

Commit:
update CHANGELOG

Commit:
merge upstream 1.20.0 into branch

tilo commented:
@donovan-lampa @dpep @corsonknowles @ldrbrandon @openbl 
upstream has merged my PR in, and i pulled the latest upstream 1.20.0 into this branch.. the new diff is tiny,
basically just additional tests we added

https://github.com/1debit/json_api_client/compare/upstream...merge-upstream-1.18.0

Commit:
update CHANGELOG

tilo commented:
places where we use our own fork:
https://github.com/search?q=org%3A1debit+1debit%2Fjson_api_client&type=code

maciej-chime commented:
do we still need to be bumping this, or should we focus on getting the remaining chime-service (and libraries?) onto the public version of the gem?

tilo commented:
at this point, we can just try to upgrade the dependent repos to the public version of `json_api_client`

https://github.com/search?q=org%3A1debit+1debit%2Fjson_api_client&type=code

gerbal committed:
⬆️ Bump Faraday dependency to v2.0

gaorlov committed:
Merge pull request #395 from sevenmind/master

Allow Faraday dependency up to v2.0

gaorlov committed:
version bump: v1.21.0

Commit:
merged upstream into this branch

tilo commented:
> wonder if we could accomplish our diff by inheriting our own resource class to do those things instead of needing to hack their library

yes, totally agree - we should not have forked this repository in the first place

corsonknowles commented:
Do you still want to merge this?
